### DISCUSSION TITLE
inverse modelling using MOOSE

### DISCUSSION THREAD

**Discussion by WilkAndy** [idaholab/moose#19086] [Q&A General] [answered]
Hi everyone,

@cpgr and i have a model containing some MaterialProperties that we'd like to estimate.  We do this by comparing the MOOSE outputs with experimental measurements.  We'd like to iteratively converge to the solution using Levenberg-Marquardt or Newton (or similar).  So, the steps are:

0. Initialize model with guestimates for the Property values
1. Run MOOSE with the current Property values, and calculate the objective function (difference between MOOSE output and experiment)
2. Modify the Property values one-by-one and Run MOOSE each time, calculating the change in the objective function. 
3. Calculate the Jacobian (change in objective function / change in Properties) and use Newton (or whatever) to update the Property values to reduce the objective function.  Goto 1, until the process converges.

We can do this process outside MOOSE, but it'd be great to use some MOOSE tools, if they exist.  Obviously (2) is embarrassingly parallel, and that's where the algorithm will spend its time.

Any suggestions?

a

**Comment by GiudGiud** [comment]
@lynnmunday @zachmprince

**Comment by lynnmunday** [comment, marked as answer]
Hi Andy,
we are working on this right now.  We have a new optimization executioner that is using TAO and the multiapp system to run the forward and adjoint problem.  We are in the process of getting this code cleaned up and made available in an external app called isopod.  I don't think it is clean enough to go into MOOSE.  I think it is going to be another few weeks before I have it documented and the tests passing and then we can make it available.

**Reply by WilkAndy** [reply]
> Hi Andy, we are working on this right now. We have a new optimization executioner that is using TAO and the multiapp system to run the forward and adjoint problem. We are in the process of getting this code cleaned up and made available in an external app called isopod. I don't think it is clean enough to go into MOOSE. I think it is going to be another few weeks before I have it documented and the tests passing and then we can make it available.

Do you think we can wait for "another few weeks", @cpgr ?   My thought is we can't.

@lynnmunday , if it's going to be publicly available, can you give us access now?

a

**Reply by zachmprince** [reply]
@WilkAndy @cpgr, I just added you to the (temporary) Isopod repository so you can check it out. There is a  material inversion example in `test/tests/formfunction/objective_minimize/material/`. This example uses a gradient-free solver, so it is pretty inefficient, but based on your suggested algorithm it doesn't sound like that is much of a problem.  A good initial guess is very important here.

**Reply by Dvora-I** [reply]
Long-time expect such a tool. Wait for isopod!

**Reply by zhang-rite** [reply]
Hope this module can go into MOOSE soon!

**Comment by cpgr** [comment]
Thanks @zachmprince and @lynnmunday. As @WilkAndy said, we probably would have had to hack something together ourselves as we are time constrained with this problem, so letting us access your work is a great help and very generous.

**Comment by WilkAndy** [comment]
I second that!  This is great - i'm just compiling isopod now.

**Comment by YaqiWang** [comment]
We had a similar situation where we want to adjust a parameter so the eigenvalue from calculation is equal to 1. We ended up writing our own solve object (or just executioner) with Newton method. The parameter must be something controllable in the input and the executioner accepts the parameter name. It will be nice to generalize this to multi-dimensional problems with the number of equations equal or greater than the number of parameters.

**Reply by zachmprince** [reply]
i would definitely like to hook Griffin in with isopod at some point. It works by computing an adjoint for the derivative in the parameter space, which is well defined for neutron transport. But it is probably overkill for something like criticality search.

**Comment by GiudGiud** [comment]
If this is a very simple 1D optimization problem, with a Real controllable parameter, you can also use the PIDTransientControl to use a PID controller to converge on the solution (as evaluated by postprocessors) during a pseudo transient
